src/schematools/loaders.py

```diff
@@ -36,45 +47,458 @@ def get_all_datasets(self) -> dict[str, DatasetSchema]:
+            dataset.tables  # noqa: ensure versioned tables are still prefetched
+            datasets[dataset_path] = dataset
+        return datasets
 
-        return dataset_schemas_from_schemas_path(self.schema_url)
 
+class FileSystemSchemaLoader(_FileBasedSchemaLoader):
+    """Loader that loads dataset schemas from the filesystem."""
 
-class URLSchemaLoader(SchemaLoader):
-    """Loader that loads dataset schemas from a url."""
+    def __init__(
+        self,
+        schema_url: Path | str,
+        *,
+        loaded_callback: Callable[[DatasetSchema], None] | None = None,
+    ):
+        """Initialize the loader with a folder where it needs to search for datasets.
+        For the convenience of importing a selected subset, it's possible
+        to point to a subfolder of the datasets repository.
+        """
+        schema_url = Path(schema_url) if isinstance(schema_url, str) else schema_url
+        if not schema_url.exists():
+            raise FileNotFoundError(schema_url)
+        if not schema_url.is_dir():
+            raise ValueError(
+                f"FileSystemSchemaLoader should receive a folder, not a file: '{schema_url}'."
+            )
 
-    def get_dataset(self, dataset_id: str, prefetch_related: bool = True) -> DatasetSchema:
-        """Gets a dataset from a url for dataset_id."""
-        from schematools.utils import dataset_schema_from_url
+        super().__init__(schema_url, loaded_callback=loaded_callback)
+        try:
+            # For compatibility with importing subfolders, this loader allows to define
+            # a subfolder as target. To avoid unexpected content in the datasets database,
+            # still try to calculate the real root so any calculated dataset-paths will work.
+            self.root = self.get_root(schema_url)
+        except ValueError:
+            # In case the real root can't be found (typically in unit tests with random layouts),
+            # assume the given folder should be treated as the root folder.
+            self.root = schema_url
 
+        # All the datasets loaded by this instance will be collected
+        # into a single cached version, so no duplicate instances are loaded.
+        self._dataset_collection = CachedSchemaLoader(self)
+
+    @classmethod
+    def from_file(cls, dataset_file: Path | str, **kwargs):
+        """Helper function to support old patterns of loading random files as schema."""
+        root = cls.get_root(dataset_file)
+        return cls(schema_url=root, **kwargs)
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self.schema_url!r})"
+
+    @classmethod
+    def get_root(cls, dataset_file: Path | str) -> Path:
+        """Resolve the real root folder.
+        This makes sure that the datasets-path value will be correct,
+        even when a subfolder is selected for importing files.
+        """
+        if isinstance(dataset_file, str):
+            dataset_file = Path(dataset_file)
```

Comment:
`Path` is idempotent (`Path(Path("foo")) == Path("foo")`), so we don't need to check:

```suggestion
        dataset_file = Path(dataset_file)
```